### ISSUE TITLE
Question about All-to-All Dispatch & Combination implementation in MOELayer

### ISSUE THREAD

**Issue by ZhiYuanZeng** [facebookresearch/fairscale#741]
## ❓ Questions and Help
Fairscale implements All-to-All Dispatch and Combination of MOE with all_to_all_single(). According to the documentation of torch.distributed, all_to_all_single splits input tensor and then scatters and concatenates the split list to all processes in a group. It can transfer a tensor with dimension (EGCM) partitioned on E devices to the tensor with dimension (GECM) partitioned on G devices (E=G), which is actually transposing a tensor partitioned on multiple devices.

However, the expert number may not equal the device number (num_local_experts!=1). In that case, why we can still use all_to_all_single for Dispatch and Combination.

**Comment by ZhiYuanZeng** [comment]
I tried a simple example to check the behavior of all_to_all_single().
```python
import torch
import os
import torch.distributed as dist
import torch.multiprocessing as mp

def setup(rank, world_size):
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '12355'

    # initialize the process group
    dist.init_process_group("nccl", rank=rank, world_size=world_size)

def cleanup():
    dist.destroy_process_group()  
  
def func(rank, world_size):
    setup(rank, world_size)
    input=torch.zeros(4).to(rank)+rank
    output = torch.empty_like(input)
    dist.all_to_all_single(output, input)
    print(output)
    cleanup()

if __name__=='__main__':
    world_size=2
    mp.spawn(func,
                args=(world_size,),
                nprocs=world_size,
                join=True)
```
![image](https://user-images.githubusercontent.com/50764547/125754439-84eeb36f-a380-426e-90b0-5bc11655d970.png)
In this example, input.shape[0]=4, however, only 2 devices is available. The output of all_to_all_single is strange, but no error occurs while running.

**Comment by min-xu-ai** [comment]
This might be a question for pytorch distributed team, but cc'ing @msbaines and @shruti-bh just in case they can chime in.

**Comment by min-xu-ai** [comment]
@ZhiYuanZeng, I briefly checked the source here: https://github.com/pytorch/pytorch/blob/fce85480b97d8d79e92e11fbcd3c03a25ae446e0/torch/distributed/distributed_c10d.py#L2480

It seems that as long as the dim 0 of the input is divisible by world_size, it would split the tensor accordingly. Here is the results when I changed the dim 0 from 4 to 6.

![image](https://user-images.githubusercontent.com/24926999/125837864-3ee0cdde-0d73-46ae-917a-fbac30c425b8.png)

So, I think this is a correct and expected behavior. Feel free to close the issue if there are no additional questions. :-)

**Comment by anj-s** [comment]
I think one useful question that @shruti-bh  can answer is how having multiple experts on a single device or less than 1 expert on a device modifies the implementation. I think this is possible but I am not clear if a simple einsum operation will be sufficient for moving input to the right expert. 

@ZhiYuanZeng Can you expound on why the output seems strange? I think it is as expected like Min mentioned.

**Comment by shruti-bh** [comment]
If there are multiple local experts, and number of experts % world_size == 0, then the logic here https://github.com/facebookresearch/fairscale/blob/master/fairscale/nn/moe/moe_layer.py#L79-L87 takes care of grouping input for all local experts that reside on a single DDP worker, and also takes care of distributing the local experts output correctly to the DDP worker that owns the respective inputs.

There is an inherent assumption that only contiguous sets of experts can be local e.g. experts 0-3 on DDP worker 0, experts 4-7 on DDP worker 1, and so on.

**Comment by ZhiYuanZeng** [comment]
Thank you for your replies. I have understood the behavior of all_to_all_single. But why not chunk before all_to_all_single, which will make the code easier to be understood. Besides, some code comments are confusing:
- https://github.com/facebookresearch/fairscale/blob/master/fairscale/nn/moe/moe_layer.py#L80, it should be GECM->EGCM?
- https://github.com/facebookresearch/fairscale/blob/master/fairscale/nn/moe/moe_layer.py#L88, it should be EGCM->GECM?

**Comment by ZhiYuanZeng** [comment]
> Thank you for your replies. I have understood the behavior of all_to_all_single. But why not chunk before all_to_all_single, which will make the code easier to be understood. Besides, some code comments are confusing:
> 
> * https://github.com/facebookresearch/fairscale/blob/master/fairscale/nn/moe/moe_layer.py#L80, it should be GECM->EGCM?
> * https://github.com/facebookresearch/fairscale/blob/master/fairscale/nn/moe/moe_layer.py#L88, it should be EGCM->GECM?
sorry, I have misunderstood these two comments:
> * https://github.com/facebookresearch/fairscale/blob/master/fairscale/nn/moe/moe_layer.py#L80
> * https://github.com/facebookresearch/fairscale/blob/master/fairscale/nn/moe/moe_layer.py#L88
they describe the dimension changes caused by reshaping operation, not that caused by  all_to_all_single. There is no more question. Thanks for your response! @min-xu-ai  @anj-s @shruti-bh

**Comment by min-xu-ai** [comment]
Thanks for letting us know and trying out fairscale! Just curious, can you share what kind of research or production work you are looking into in using the moe layer?
